dssi/doc/dssi_list_plugins.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH dssi_list_plugins 1 "June 23rd, 2010"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
dssi_list_plugins \- list available DSSI plugins
.SH SYNOPSIS
.B dssi_list_plugins
[-v|--verbose]
.SH DESCRIPTION
.B dssi_list_plugins
scans a list of directories specified by the environment variable
.BR DSSI_PATH ,
looking for DSSI (or LADSPA) shared libraries.  It reports each library
found, and lists the plugins contained within the library. It ignores
subdirectories and libtool *.la files; otherwise it will complain
if any of the files found are not DSSI or LADSPA shared libraries.
.SH OPTIONS
.TP
.PD 0
.B -v
.TP
.PD
.B --verbose
Causes
.B dssi_list_plugins
to be more verbose in its reporting. It will note conditions which
are typically ignored by most DSSI hosts, yet may indicate
problems, including malformed or unreadable components of
.BR DSSI_PATH ,
directories which contain no plugin libraries, libraries which appear
more than once in the
.BR DSSI_PATH ,
and libraries which contain no plugins.
.SH ENVIRONMENT
.TP
.B DSSI_PATH
A colon-separated list of directories to scan for DSSI plugins.
.SH "CONFORMING TO"
DSSI RFC version 1.0.
.SH SEE ALSO
.BR dssi_analyse_plugin (1),
http://dssi.sourceforge.net/


libmodsynth/ChangeLog
20111006 CVS commit cannam

 * Fixed crash in less_trivial_synth.c in init() with gcc 4.6

20110523 CVS commit smbolton

 * Fixed linking with GNU gold linker (thanks to Alessio Treglia for
     the bug report).

 * Use RTLD_LOCAL when dlopen()ing plugins.

20100919 CVS commit and 1.1.0 release, smbolton

 * Man page updates.

20100916 CVS commit cannam

 * Added '-c' option for client name to jack-dssi-host.

20100628 CVS commit smbolton

 * Fixed configure.ac to check for OS X Qt 4 frameworks, and to add
     '-lX11' to QT_LIBS when Qt is X11-based.

20100627 CVS commit smbolton

 * Added dssi_list_plugins and dssi_analyse_plugin tools.

 * Updated README for Qt 4 and new tools.

 * Fixed jack-dssi-host.c ALSA client ports to be of type
     'application'.

 * Fixed trivial_synth.c getControllerTS() to use DSSI_CC().

 * examples/Makefile.am fixes: reverted trivial_sampler_qt_CXXFLAGS
     to again include $(ALSA_CFLAGS); added math lib to
     trivial_synth_la_LIBADD.

 * Version bump to 1.0.1.

20100601 CVS commit smbolton

 * Updated example GUIs to use Qt 4 (thanks to Charles Flèche!)

20100530 CVS commit smbolton

 * Fixed an uninitialized variable bug in less_trivial_synth.c.

20090109 CVS commit and 1.0.0 release, smbolton

 * Added manpages (thanks to Mark Hymers and Debian!)
 * Freshening URLs, etc., for 2009.
 * Small dssi_osc_update fixes.

20081214 CVS commit smbolton

 * Implemented the 'sample-rate' OSC method in both jack-dssi-host
     and the example plugins, and described it in RFC.txt. 

20081212 CVS commit smbolton

 * jack-dssi-host: changed to use jack_client_open() instead of
     deprecated jack_client_new() (requires jack >=0.99.14).

 * jack-dssi-host: fixed audio output port connection to avoid
     JACK MIDI ports.

 * Example plugins: use gcc/glibc 'constructor' and 'destructor'
     attributes instead of the deprecated _init() and _fini()
     functions (needed for OS X 10.4 'Tiger').

 * Multiple clarifications in RFC.txt: configure, program, and
     control OSC messages are not echoed back to the originator;
     argv[0] supplied to the UI should include the full path (all
     existing hosts do this, to the best of my knowledge);
     eliminated some inconsistency between use of 'PLUGINS' and
     'MYPLUGINS' in examples.

20061213 CVS commit cannam

 * Add another simple example -- the basic Karplus and Strong synth.

20051012 CVS commit and 0.9.1 release, smbolton

 * Updated examples/dssi_osc_update.c so it works again.

 * jack-dssi-host: upstreamed gcc4 fix from Gentoo ebuild.

20051006 CVS commit smbolton

 * Added an '-a' option to jack-dssi-host, to disable automatic
     connection of audio output ports to JACK physical output ports.

 * Updated why-use.txt to reflect currently available plugins.

20050723 CVS commit cannam

 * jack-dssi-host: send updates out to UI if received from
     elsewhere.

20050525 CVS commit smbolton

 * Version bump to 0.9.1.

 * Replaced static makefiles with a full autoconf/automake/libtool
     build system.  Hope it solves more problems than it creates....

 * Dropped message_buffer.[ch] use from less_trivial_synth.c, since
     it's been commented out, and crashes some hosts.

 * Conditionally omit the ALSA sequencer client code from
     jack-dssi-host.c if building on a system without ALSA.

 * Updated documentation to reflect FluidSynth-DSSI move and
     autotools build system.

20050516 CVS commit smbolton

 * Moved FluidSynth-DSSI to its own separate CVS module.

Summary of changes, version 0.9 (20041103) through 20050510

 * Numerous clarifications in the documentation.

 * Fixed less_trivial_synth.c for use on big-endian systems.

 * jack-dssi-host: better handling of LADSPA hinting, support for
     audio inputs, support for LADSPA-only plugins, better signal
     handling with cleaner shutdown, several smaller bug fixes.



dssi/examples/dssi_list_plugins.c
/* dssi_list_plugins.c
 *
 * Written by Sean Bolton, with inspiration from Richard Furse's listplugins
 * from the LADSPA SDK.
 *
 * This program is in the public domain.
 *
 * $Id: dssi_list_plugins.c,v 1.1 2010/06/27 19:30:09 smbolton Exp $
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <dirent.h>
#include <errno.h>
#include <sys/stat.h>
#include <dlfcn.h>

#include <ladspa.h>
#include "dssi.h"

struct duplicate_list {
    struct duplicate_list *next;
    const char *item;
};

struct duplicate_list *dup_list = NULL;

int verbose = 0;

void
check_for_duplicates(const char *filename)
{
    struct duplicate_list *d;

    for (d = dup_list; d; d = d->next) {
        if (!strcmp(d->item, filename)) {
            printf("-- warning: shared library '%s' appears more than once on DSSI_PATH\n", filename);
            return;
        }
    }

    /* not found, add it to the list */
    d = (struct duplicate_list *)malloc(sizeof(struct duplicate_list));
    d->item = strdup(filename);
    d->next = dup_list;
    dup_list = d;
}

void
free_dup_list(void)
{
    struct duplicate_list *d;

    while (dup_list) {
        d = dup_list;
        dup_list = d->next;
        free((char *)d->item);
        free(d);
    }
}

void
list_dssi_plugins(DSSI_Descriptor_Function descriptor_function)
{
    int i;
    const DSSI_Descriptor *descriptor;

    for (i = 0; (descriptor = descriptor_function(i)); i++)
        printf("\t%-16s  %s\n", descriptor->LADSPA_Plugin->Label, 
                                descriptor->LADSPA_Plugin->Name);

    if (verbose && i == 0)
        printf("-- odd ... no plugins found within this shared object\n");
}

void
list_ladspa_plugins(LADSPA_Descriptor_Function descriptor_function)
{
    int i;
    const LADSPA_Descriptor *descriptor;

    for (i = 0; (descriptor = descriptor_function(i)); i++)
        printf("\t%-16s  (%lu) %s\n", descriptor->Label, descriptor->UniqueID, descriptor->Name);

    if (verbose && i == 0)
        printf("-- odd ... no plugins found within this shared object\n");
}

void
list_directory(char *directory)
{
    int directory_length = strlen(directory);
    DIR *dir;
    struct dirent *dirent;
    char *filename;
    struct stat statbuf;
    void *handle;
    int found = 0;

    if (directory_length == 0) {
        if (verbose) printf("-- warning: skipping zero-length element in DSSI_PATH\n");
        return;
    }
    if (directory[directory_length - 1] == '/')
        directory[directory_length - 1] = 0; /* slash gets added again below */

    dir = opendir(directory);
    if (!dir) {
        if (verbose) printf("-- warning: couldn't open DSSI_PATH directory element '%s'\n", directory);
        return;
    }

    if (verbose) printf("-- scanning directory %s\n", directory);

    while ((dirent = readdir(dir)) != NULL) {

        if (!strcmp(dirent->d_name, ".") || !strcmp(dirent->d_name, ".."))
            continue;

        filename = malloc(directory_length + strlen(dirent->d_name) + 2);
        sprintf(filename, "%s/%s", directory, dirent->d_name);

        if (stat(filename, &statbuf)) {
            if (verbose)
                printf("-- warning: couldn't stat file '%s': %s\n", filename, strerror(errno));
            free(filename);
            continue;
        }
        if (S_ISDIR(statbuf.st_mode)) { /* silently skip subdirectories */
            free(filename);
            continue;
        }

        handle = dlopen(filename, RTLD_LAZY);
        if (handle) {
            DSSI_Descriptor_Function dssi_descriptor_function;
            LADSPA_Descriptor_Function ladspa_descriptor_function;

            if (verbose)
                check_for_duplicates(dirent->d_name);

            dssi_descriptor_function = (DSSI_Descriptor_Function)dlsym(handle, "dssi_descriptor");
            ladspa_descriptor_function = (LADSPA_Descriptor_Function)dlsym(handle, "ladspa_descriptor");

            if (dssi_descriptor_function) {
                printf("%s\n", filename);
                list_dssi_plugins(dssi_descriptor_function);
                found++;
            } else if (ladspa_descriptor_function) {
                printf("%s (LADSPA-only)\n", filename);
                list_ladspa_plugins(ladspa_descriptor_function);
            } else {
                if (verbose)
                    printf("-- warning: shared object '%s' is neither a LADSPA nor DSSI plugin\n", filename);
            }
            dlclose(handle);
        } else { /* not a library */
            int len = strlen(dirent->d_name);

            if (len > 4 && !strcmp(dirent->d_name + len - 3, ".la")) {  /* libtool *.la file */
                /* if (verbose) printf("-- skipping file '%s'\n", filename); */
            } else
                if (verbose)
                    printf("-- warning: couldn't dlopen file '%s': %s\n", filename, dlerror());
        }
        free(filename);
    }

    closedir(dir);

    if (verbose && found == 0)
        printf("-- odd ... no DSSI plugins were found in this directory\n");
}

void
usage(const char *program_name)
{
    fprintf(stderr, "usage: %s [-v|--verbose]\n", program_name);
    fprintf(stderr, "Scan the directories listed in environment variable DSSI_PATH, and list\n");
    fprintf(stderr, "the DSSI plugins found therein.  Optional arguments:\n");
    fprintf(stderr, "  -v, --verbose    describe the scan and error conditions more verbosely.\n");

    exit(1);
}

int 
main(int argc, char *argv[])
{
    char *path, *pathtmp, *element;

    while (argc > 1) {
        argc--;
        if (!strcmp(argv[argc], "-v") || !strcmp(argv[argc], "--verbose"))
            verbose = 1;
        else
            usage(argv[0]); /* does not return */
    }

    path = getenv("DSSI_PATH");
    if (!path) {
        path = "/usr/local/lib/dssi:/usr/lib/dssi";
        fprintf(stderr, "warning: DSSI_PATH not set, defaulting to '%s'\n", path);
    }

    path = strdup(path);
    pathtmp = path;
    while ((element = strtok(pathtmp, ":")) != 0) {
        pathtmp = NULL;
        list_directory(element);
    }
    free(path);

    free_dup_list();

    return 0;
}

